src/editor_transport.rs

```diff
@@ -127,8 +127,13 @@ pub fn start_unix(path: &path::Path, sender: Sender<EditorRequest>) {
                             continue;
                         }
                         debug!("From editor: {}", request);
-                        let request: EditorRequest =
-                            toml::from_str(&request).expect("Failed to parse editor request");
+                        let request: EditorRequest = match toml::from_str(&request) {
+                            Ok(req) => req,
+                            Err(err) => {
+                                error!("Failed to parse editor request: {}", err);
```

Comment:
I'm not convinced this is an improvement.
Unlike in the previous commit, the user has to actively create a bad configuration.
I feel like it's better to fail noisily and force them to fix it instead of ignoring the configuration (users might not read the log until later).

I think it would be okay to avoid panics if we send an `lsp-show-error` command here, to tell the user.

(Optionally, we could run `lsp-show-error` only if the wrong configuration applies to the language in context.
However, deducing that from invalid TOML input sounds like very high effort for very low gain.)

BTW `lsp-server-configuration` is deprecated in favor of `lsp-config`.

Reply:
I wasn't sure how to know which client to notify if we can't parse the toml. Is there an easy way for a user to realize that `kak-lsp` panicked if they aren't manually running it in a separate terminal? As far as I can tell, it just silently restarts in the background with no indication of what's happening. Maybe there's a quick and dirty solution in grepping the input for  `client = "..."`.

Or, of course, this crash could probably be mitigated enough by fixing any remaining shell escaping errors. I just wonder if there are any spooky edge cases that might remain, e.g. if it's possible for a command to be half-written if the user ctrl+c's at the wrong time, or their terminal emulator crashes, etc.

Reply:
> Is there an easy way for a user to realize that `kak-lsp` panicked if they aren't manually running it in a separate terminal?

Yeah, sadly there is no easy way other than looking at the output (or log file).
I want to make something like neovim's :LspInfo that shows debug info about all running servers.
  (Unfortunately the "all" bit doesn't seem straightforward: I might be wrong but I think kak-lsp will spawn a separate controller process for each filetype, so the languages don't share any state; there is no central list of all controllers.)
  Still, we can write an `:lsp` command (removing the old one) that shows the status of the current buffer's server in a scratch buffer.
  It should display the last few lines of the log.

---

Normally I'd rather fix the escaping straight away. Probably not worth the effort, here's why:
We serialize it like

    value="$(printf %s "$value"|sed -e 's/\\=/=/g')"
    printf '%s = %s\n' "$quotedkey" "$value"

so we don't add quotes to value, because it could be any of bool/string/number, etc.
We escape it (only if it's a string), but that might change the behavior for users who already worked around this bug and escaped the value themselves.

We should encourage users to migrate to lsp_config, which uses real TOML syntax (and uses lsp-show-error if it's invalid). How did you find lsp_server_configuration? It should have been removed from docs, the option is hidden. I guess you already used it before?

> Maybe there's a quick and dirty solution in grepping the input for  `client = "..."`.

Fine by me. Alternatively, sending a message to all clients would also work, but that's hardly easier.

> I just wonder if there are any spooky edge cases that might remain, e.g. if it's possible for a command to be half-written if the user ctrl+c's at the wrong time, or their terminal emulator crashes, etc.

Possible but unlikely. When kak reads ctrl+c, it sends SIGINT to the entire process group, which includes the `kak-lsp --request` process (but not the controller process which is daemonized). But `kak-lsp --request`is so short-lived, and the usual TOML payload easily fits in one write(2) syscall, which won't be interrupted even by signals.

Reply:
> > Is there an easy way for a user to realize that `kak-lsp` panicked if they aren't manually running it in a separate terminal?
> 
> Yeah, sadly there is no easy way other than looking at the output (or log file). I want to make something like neovim's :LspInfo that shows debug info about all running servers. (Unfortunately the "all" bit doesn't seem straightforward: I might be wrong but I think kak-lsp will spawn a separate controller process for each filetype, so the languages don't share any state; there is no central list of all controllers.) Still, we can write an `:lsp` command (removing the old one) that shows the status of the current buffer's server in a scratch buffer. It should display the last few lines of the log.

I don't completely understand why panicking is better than logging an error in the failed-to-parse case, if it's as hard for regular users to notice either way. But sounds like the debug info command is the right way forward.

Fun fact: I managed to get failed-to-parse errors from editing files with some bad unicode characters (mostly scratch text files I paste things into), whereas e.g. vscode seems to forward broken files like this to its language servers to complain about instead. I guess the buffer contents would need to be base64'd to allow passing raw bytes through toml.

```
Aug 07 07:27:25.367 ERRO Failed to parse editor request: invalid character in string: `\u{1}` at line 5013 column 18, module: kak_lsp::editor_transport:133
```

> We should encourage users to migrate to lsp_config, which uses real TOML syntax (and uses lsp-show-error if it's invalid). How did you find lsp_server_configuration? It should have been removed from docs, the option is hidden. I guess you already used it before?

Looks like it's from when I first configured kak-lsp... About 3 years ago, haha.

> > Maybe there's a quick and dirty solution in grepping the input for  `client = "..."`.
> 
> Fine by me. Alternatively, sending a message to all clients would also work, but that's hardly easier.

I implemented this, with an extra regex for `hook = true`. Feels a bit hacky, but IMO better than crashing.